src/librustc_back/target/mod.rs

```diff
@@ -578,17 +588,61 @@ impl Target {
                         .map(|s| s.to_string() );
                 }
             } );
+            ($key_name:ident, LinkerFlavor) => ( {
+                let name = (stringify!($key_name)).replace("_", "-");
+                obj.find(&name[..]).and_then(|o| o.as_string().and_then(|s| {
+                    match s {
+                        "em" => base.options.$key_name = LinkerFlavor::Em,
+                        "gnu" => base.options.$key_name = LinkerFlavor::Gnu,
+                        "ld" => base.options.$key_name = LinkerFlavor::Ld,
+                        "msvc" => base.options.$key_name = LinkerFlavor::Msvc,
+                        _ => return Some(Err(format!("'{}' is not a valid value for \
+                                                      linker-flavor. Use 'em', 'gnu', 'ld' \
+                                                      or 'msvc.",
+                                                     s))),
+                    }
+                    Some(Ok(()))
+                })).unwrap_or(Ok(()))
+            } );
+            ($key_name:ident, link_args) => ( {
+                let name = (stringify!($key_name)).replace("_", "-");
+                if let Some(obj) = obj.find(&name[..]).and_then(|o| o.as_object()) {
+                    let mut args = LinkArgs::new();
+                    for (k, v) in obj {
+                        let k = match &k[..] {
+                            "em" => LinkerFlavor::Em,
+                            "gnu" => LinkerFlavor::Gnu,
+                            "ld" => LinkerFlavor::Ld,
+                            "msvc" => LinkerFlavor::Msvc,
```

Comment:
These same four strings have been repeated quite a few times throughout this PR, perhaps they could be centralized to one location though?